developer/python-django-devtools-reference.rst

```diff
@@ -0,0 +1,229 @@
+tools according to programmer’s demands that greatly reduces development
+time. You can find some more details with examples and tutorials
+`on the official DRF website <https://www.django-rest-framework.org>`_
+or consider checking out the DRF tutorial along with some other tutorials
+on `this website <https://wsvincent.com/>`_.
+
+Django Extensions
+-----------------
+
+`Django Extensions <https://django-extensions.readthedocs.io/>`_ is a
+collection of custom extensions for the Django Framework. These include
+management commands, additional database fields, admin extensions and
+much more but we will focus more on 2 of them for now, namely
+``shell_plus`` and ``runserver_plus``.
+
+Django Extensions can be installed with:
+
+``pip install django-extensions``
+
+`shell_plus
+<django-extensions.readthedocs.io/en/latest/shell_plus.html>`_:
+Django shell with autoloading of the apps database models and subclasses
+of user-defined classes.
+
+`runserver_plus
+<django-extensions.readthedocs.io/en/latest/runserver_plus.html>`_:
+Typical runserver with Werkzeug debugger baked in.
+
+Django Debug Toolbar
+--------------------
+
+The Django Debug Toolbar is a configurable set of panels that display
+various debug information about the current request/response and when
+clicked, display more details about the panel’s content.
+
+It can be installed with:
+
+``pip install django-debug-toolbar``
+
+More details can be found on its `website
+<https://django-debug-toolbar.readthedocs.io/en/latest/>`_
+or on `pypi <https://pypi.org/project/django-debug-toolbar/>`_.
+
+IPython and ipdb
+----------------
+
+IPython (Interactive Python) is a command shell for interactive computing
+in multiple programming languages, originally developed for the Python
+programming language, that offers introspection, rich media, shell
+syntax, tab completion, and history.
+
+It provides:
+- A powerful interactive shell.
+- A browser-based notebook interface with support for code, text,
+mathematical expressions, inline plots and other media.
+- Support for interactive data visualization and use of GUI toolkits.
+- Flexible, embeddable interpreters to load into one's own projects.
+- Tools for parallel computing.
+
+More details including installation and updates can be found on the
+`official webite <https://ipython.org>`_ and a simple usage tutorial
+can be found `here <http://cs231n.github.io/ipython-tutorial/>`_.
+
+As for ipdb, it exports functions to access the IPython debugger. It
+comes along with ipython. More details about ipdb can be obtained
+on `its pypi page <https://pypi.org/project/ipdb/>`_.
+
+How to enable these dev tools in OpenWISP modules
+-------------------------------------------------
+
+These tools can be implemented in OpenWISP modules and can significantly
+improve the workflow. Let's do a walkthrough of how to use them in
+`OpenWISP Controller <https://github.com/openwisp/openwisp-controller>`_
```

Comment:
did you already mention that the same can be done for other modules?

Reply:
yep. L172